File: src/year2024/day1/day1.lua
#!/usr/bin/env lua
-- [Day 1: Historian Hysteria](https://adventofcode.com/2024/day/1)

input_file = arg[1] or "input.txt"

local left = {}
local right = {}

for line in io.lines(input_file) do
    local a, b = line:match("(%d+)%s*(%d+)")
    table.insert(left, tonumber(a))
    table.insert(right, tonumber(b))
end

-- Sort both lists
table.sort(left)
table.sort(right)

-- Part 1: Calculate the sum of absolute differences
local part1 = 0
for i = 1, #left do
    part1 = part1 + math.abs(left[i] - right[i])
end
print(part1)

-- Part 2: Count occurrences of each element in the right list
local right_counts = {}
for _, num in ipairs(right) do
    right_counts[num] = (right_counts[num] or 0) + 1
end

-- Calculate the weighted sum for part 2
local part2 = 0
for _, num in ipairs(left) do
    part2 = part2 + num * (right_counts[num] or 0)
end
print(part2)

File: src/year2024/day1/day1.java
// [Day 1: Historian Hysteria](https://adventofcode.com/2024/day/1)

// javac day1.java
// java day1 [input.txt]

import java.io.*;
import java.nio.file.*;
import java.util.*;

class day1 {
    public static void main(String[] args) throws IOException {
        // Default to "input.txt" if no arguments are passed
        String inputFile = args.length == 0 ? "input.txt" : args[0];

        // Read the file into two lists: left and right
        List<Integer> left = new ArrayList<>();
        List<Integer> right = new ArrayList<>();

        // Read the file and split the numbers
        Files.lines(Paths.get(inputFile)).forEach(line -> {
            String[] parts = line.split("\\s+");
            left.add(Integer.parseInt(parts[0]));
            right.add(Integer.parseInt(parts[1]));
        });

        // Sort both lists
        Collections.sort(left);
        Collections.sort(right);

        // Part 1: Calculate the sum of absolute differences
        int part1 = 0;
        for (int i = 0; i < left.size(); i++) {
            part1 += Math.abs(left.get(i) - right.get(i));
        }
        System.out.println(part1);

        // Part 2: Count the occurrences of each number in the right list
        Map<Integer, Integer> rightCounts = new HashMap<>();
        for (int num : right) {
            rightCounts.put(num, rightCounts.getOrDefault(num, 0) + 1);
        }

        // Calculate the weighted sum of left elements based on right counts
        int part2 = 0;
        for (int num : left) {
            part2 += num * rightCounts.getOrDefault(num, 0);
        }
        System.out.println(part2);
    }
}
